app/js/RedisManager.js

```diff
@@ -49,6 +49,55 @@ const historyKeys = Settings.redis.history.key_schema // note: this is track cha
 module.exports = RedisManager = {
   rclient,
 
+  serializeDocCore(
+    jsonLines,
+    hash,
+    jsonRanges,
+    projectId,
+    pathName,
+    projectHistoryId
+  ) {
+    return [
+      jsonLines,
+      hash,
+      jsonRanges,
+      projectId,
+      JSON.stringify(pathName),
+      String(projectHistoryId)
+    ].join('\u0000')
+  },
+
+  deserializeDocCore(blob) {
+    const [
+      docLines,
+      hash,
+      ranges,
+      projectId,
+      pathName,
+      projectHistoryId
+    ] = blob.split('\u0000')
```

Comment:
was the main reason for using null bytes to prevent having JSON in JSON with?

Reply:
> was the main reason for using null bytes to prevent having JSON in JSON with?

Yes.

The serialization of 1) doc lines and 2) ranges is done separately for 1) hashing doc lines and 2) limiting the size of all ranges. The hashing could be adjusted to cover the full blob (doc lines, ranges, ...) and limiting the size could be adjusted to check the full blob size (blobLength - maxDocLength < maxRangesSize, which is slightly lossy but may be OK or not). We could then do one JSON serialization, which may simplify things overall.
Another alternative to the size check could be a custom JSON.stringify function:
```js
const details = { docLines, ranges, ... }
JSON.stringify(details, (key, value) => {
  if (this === details && key === 'ranges') {
    if (value.length > maxRangesLength) {
      throw new Error('ranges size exceeds limit')
    }
  }
  return value
})
```